### PR TITLE
use dts for mmu mapping

### PR DESCRIPTION
**Pull request by JiafeiPan**
devicetree: add mmu mapping support

   Currently for ARM64, mmu mapping entries are hardcoded in soc's
   mmu_regions.c, and such mapping will be used for all board and
   applications although some mapping are not useful for some
   applications.

   So this patch provides another method to add mmu mapping entry
   by devicetree, so such each application can specify mmu mapping
   by devicetree overlay.

   Devicetree need to do the following update:
    1. Add property "zephyr,memory-region" to specify mmu entry name.
    2. Add property "zephyr,memory-region-mmu" to specify mmu entry
       attribute with string array which is valid c token from  architecture
        speficied MMU header file.
    3. The address and size will from "reg" node.

    For example:

            ccm: ccm@30380000 {
                    compatible = "nxp,imx-ccm";
                    reg = <0x30380000 DT_SIZE_K(64)>;
                    zephyr,memory-region = "CCM";
                    zephyr,memory-region-mmu = "MT_DEVICE_nGnRnE", "MT_P_RW_U_NA", "MT_NS";
                    #clock-cells = <3>;
           };

    And then in C code to include mmu entries specified in devicetree:
    For example in "mmu_regions.c" for ARM64 platform
    ...
    ...

    static const struct arm_mmu_region mmu_regions[] = {
            ...
            LINKER_DT_REGION_MMU(MMU_REGION_FLAT_ENTRY)
            ...
    };

With this dts example, the code generated will be:
```
   static const struct arm_mpu_region mpu_regions[] = {
       ...
       { .name = "CCM",
         .base = 0x30380000,
         .size = DT_SIZE_K(64),
         .attr = MT_DEVICE_nGnRnE | MT_P_RW_U_NA | MT_NS, },
       ...
   };
```

### CODE REVIEW

**Comment by JiafeiPan** [comment]
The similar design for MMU with https://github.com/zephyrproject-rtos/zephyr/pull/43119/files

**Comment by carlocaione** [comment]
@JiafeiPan did you mean to close this PR?

**Comment by JiafeiPan** [comment]
> @JiafeiPan did you mean to close this PR?

Hi, @carlocaione , my initial idea is to leverage compatible "zephyr,memory-region" to figure out the dts node which need to do mmu mapping, so I move zephyr,memory-region.yaml to be included in base.yaml, but I found with this PR, there is some error

**Comment by JiafeiPan** [comment]
> > @JiafeiPan did you mean to close this PR?
> 
> Hi, @carlocaione , my initial idea is to leverage compatible "zephyr,memory-region" to figure out the dts node which need to do mmu mapping, so I move zephyr,memory-region.yaml to be included in base.yaml, but I found with this PR, there is some error for ARM32 platform, so I want to change the design to only to do mmu mapping the dts node with property of "zephyr,memory-region-mmu" but not use compatible string "zephyr,memory-region", but currently I can't find any API which is like "DT_FOREACH_STATUS_OKAY_WITH_PROPERTY(property_string, FN, ...)", any suggestions? thanks.

Hi, @carlocaione , I worked out a new implementation, it will don't use "zephyr,memory-region"  compatible, only need to check the node: 
1. status okay
2. both with property: "zephyr,memory-region",       "zephyr,memory-region-mmu",  "reg"

Then add mmu mapping entry for these dts nodes. please help to review, thanks.

**Comment by JiafeiPan** [comment]
Dear maintainers, this PR is ready for review, please help to review, and I would like to merge  it into Zephyr v3.2 release,  thanks.

**Comment by carlocaione** [comment]
What's the use case here?

The problem I have with this PR is that we should try to avoid and discourage the usage of `mmu_regions.c` entirely. That was a quick and dirty way to be able to have a 1:1 (direct) mapping in the MMU when the MMU driver was in its infancy. Now the correct way for a driver to have its MMIO space mapped is by using the Device MMIO API.

In your example the CCM driver should be using the Device MMIO API to map its own space without having to resort to extend the `mmu_regions.c` at all. If an application doesn't need that driver (and its MMU mapping) then it can simply remove it from the DT (and the mapping will be gone with it).

The only reason `mmu_regions.c` is still there is because the GIC driver is not a proper driver and cannot use the Device MMIO APIs (yet).

So, from a technical prospective this PR is fine but I'm worried this will basically enable the bad habit of using `mmu_regions.c` instead of modifying the drivers to properly support the Device MMIO APIs.


See #46477 for an extensive discussion on this topic.

@povergoing @SgrrZhf what do you think about this?

**Comment by povergoing** [comment]
> Now the correct way for a driver to have its MMIO space mapped is by using the Device MMIO API.

Agreed, 1) it should be the drivers' responsibility to do the MMIO mapping and then do some configs or something important. 2) using DTS for mapping seems to increase the complexity of device tree usage. I doubt it could be too many config entries in the DTS if we apply this for a complicated device like PCI + AHCI

**Comment by JiafeiPan** [comment]
I agree with that driver should take responsibility to do the MMIO mapping.

But Carlo said "the big issue with Zephyr is that 95% of the drivers are not using the device MMIO API",  especially in some cases user maybe use hal driver directly which is not a standard Zephyr driver, and in general some hal driver will be shared with other RTOS, so not possible to modify hal driver to add MMIO mapping.

Another concern is some hal driver is using physical address directly, if we don't do 1:1 mapping, hal driver should has memory access issues.

So my suggestions is we need to ask drivers to support MMIO mapping if it is possible, in the other way, provide user another easier method to do 1:1 mapping.

Thanks.

File: include/zephyr/devicetree.h
Comment (mbolivar-nordic) [inline include/zephyr/devicetree.h:2107]: 1. Rename this to DT_FOREACH_STATUS_OKAY_NODE_VARGS. It is the "VARGS" version of DT_FOREACH_STATUS_OKAY_NODE. It is not the "NODE" version of DT_FOREACH_STATUS_OKAY_VARGS.
2. Add test coverage in tests/lib/devicetree/api/src/main.c

File: dts/bindings/base/base.yaml
Comment (mbolivar-nordic) [inline dts/bindings/base/base.yaml:129]: Whoa, hold on, why are these going into base.yaml? The commit log doesn't explain. Adding base.yaml properties needs a good reason.

File: scripts/dts/gen_defines.py
Comment (mbolivar-nordic) [inline scripts/dts/gen_defines.py:865]: You must update doc/guides/dts/macros.bnf whenever you add new generated macros.

File: include/zephyr/linker/devicetree_regions.h
Comment (mbolivar-nordic) [inline include/zephyr/linker/devicetree_regions.h:217]: Is it really worth adding UTIL_AND_AND just to use it here? Not a blocking comment since I don't maintain the util header, but it doesn't seem justified since you can just nest COND_CODE_1 here. I think it is risky to start adding macros like UTIL_AND_AND -- where do you stop? UTIL_AND_OR? UTIL_OR_AND? UTIL_OR_OR? UTIL_AND_AND_AND? ...

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]

File: dts/bindings/base/base.yaml
Comment (JiafeiPan) [reply dts/bindings/base/base.yaml:129]: The idea is that these two prosperities can be used in any node which has "reg" property,  as "reg" is is base.yaml, so I think we also need to put these two together with "reg",  otherwise if they are put in some standalone yaml file, for example, put them into dts/bindings/base/zephyr,memory-region.yaml where current "zephyr, memory-region" locates, we have to include this ymal file in each device driver node yaml files, I don't think it is a better choice, so I suggest put it base.yaml, so that no need update the other yaml files anymore.

File: include/zephyr/devicetree.h
Comment (JiafeiPan) [reply include/zephyr/devicetree.h:2107]: okay, will update it.

File: scripts/dts/gen_defines.py
Comment (JiafeiPan) [reply scripts/dts/gen_defines.py:865]: okay, will update it.

File: include/zephyr/linker/devicetree_regions.h
Comment (JiafeiPan) [reply include/zephyr/linker/devicetree_regions.h:217]: Okay, will remove this macro.

**Comment by carlocaione** [comment]
> especially in some cases user maybe use hal driver directly which is not a standard Zephyr driver, and in general some hal driver will be shared with other RTOS, so not possible to modify hal driver to add MMIO mapping.

> Another concern is some hal driver is using physical address directly, if we don't do 1:1 mapping, hal driver should has memory access issues.

Those are valid concerns, still an hard pillow to swallow TBH.

File: dts/bindings/base/base.yaml
Comment (carlocaione) [inline dts/bindings/base/base.yaml:118]: This is actually only true for architectures supporting this. ARM64 is not one of those. Can you expand this with:

`Signify that this node should result in a dedicated linker script memory region in the final executable when supported by the linker script`.

File: include/zephyr/linker/devicetree_regions.h
Comment (carlocaione) [inline include/zephyr/linker/devicetree_regions.h:216]: Can you rename the current `_CHECK_ATTR_FN` to `_CHECK_MPU_ATTR_FN` for coherence as well?

File: include/zephyr/linker/devicetree_regions.h
Comment (carlocaione) [inline include/zephyr/linker/devicetree_regions.h:228]: same here, can you rename `_CHECK_APPLY_FN` to `_CHECK_APPLY_MPU_FN`?

File: include/zephyr/linker/devicetree_regions.h
Comment (carlocaione) [inline include/zephyr/linker/devicetree_regions.h:367]: This between `@code{.dts}` and `@endcode`

While you are at it, can you fix lines 250-257 in the same file?

File: include/zephyr/linker/devicetree_regions.h
Comment (carlocaione) [inline include/zephyr/linker/devicetree_regions.h:407]: At this point I'm realizing how misnamed are `LINKER_DT_REGION_MPU` and this macro  because these macros have nothing to do with the linker script.

Can you rename those to something like `BUILD_DT_REGION_{MPU,MMU}`?

File: soc/arm64/nxp_imx/mimx8m/mmu_regions.c
Comment (carlocaione) [inline soc/arm64/nxp_imx/mimx8m/mmu_regions.c:39]: This should actually be added to all the `mmu_regions.c` files for the ARM64 arch.

**Review by carlocaione — CHANGES_REQUESTED** [review]
Besides the comments left, if we decide to have this merged we should at lest convert all the current `mmu_regions.c` files to use this new mechanism and at least having the GIC entries removed from there.

**Comment by SgrrZhf** [comment]
> Now the correct way for a driver to have its MMIO space mapped is by using the Device MMIO API.

I think so. If we really want a 1:1 mapping, maybe we could introduce a feature in Device MMIO API that does 1:1 mapping. It's that possible?

**Comment by SgrrZhf** [comment]
Understood, those memory pages must be 1:1 mapped before running some HAL driver.

File: dts/bindings/base/base.yaml
Comment (JiafeiPan) [reply dts/bindings/base/base.yaml:118]: Updated, but replace linker with build.

File: include/zephyr/linker/devicetree_regions.h
Comment (JiafeiPan) [reply include/zephyr/linker/devicetree_regions.h:367]: Done.

Comment (JiafeiPan) [reply include/zephyr/linker/devicetree_regions.h:367]: Done.

File: include/zephyr/linker/devicetree_regions.h
Comment (JiafeiPan) [reply include/zephyr/linker/devicetree_regions.h:407]: Done.

File: soc/arm64/nxp_imx/mimx8m/mmu_regions.c
Comment (JiafeiPan) [reply soc/arm64/nxp_imx/mimx8m/mmu_regions.c:39]: Done

File: include/zephyr/linker/devicetree_regions.h
Comment (JiafeiPan) [reply include/zephyr/linker/devicetree_regions.h:216]: Done.

**Comment by JiafeiPan** [comment]
Dear maintainers, any more comments or can we merge this PR by today? thanks.

**Comment by carlocaione** [comment]
> Dear maintainers, any more comments or can we merge this PR by today? thanks.

At least @mbolivar-nordic needs to ACK this before merging

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I  think either there should be a good reason why all of these properties belong in base.yaml, or they all should continue to live in zephyr,memory-region.yaml. The rest of the DT API changes look good, though, thank you.

**Comment by galak** [comment]
Need to read through the thread, but this really feels like it should be in the driver and NOT in devicetree.

**Comment by JiafeiPan** [comment]
> I think either there should be a good reason why all of these properties belong in base.yaml, or they all should continue to live in zephyr,memory-region.yaml. The rest of the DT API changes look good, though, thank you.

Hi, @mbolivar-nordic, The idea is that these two properties can be used in any node which has "reg" property, currently "reg" is is base.yaml, so I think we'd better put these two together with "reg", otherwise if they are put in some standalone yaml file, for example, put them into dts/bindings/base/zephyr,memory-region.yaml where current "zephyr, memory-region" locates, we have to include this ymal file in each device driver node yaml files which want to use this feature, there will be more modifications.

**Comment by JiafeiPan** [comment]
> I think either there should be a good reason why all of these properties belong in base.yaml, or they all should continue to live in zephyr,memory-region.yaml. The rest of the DT API changes look good, though, thank you.

Hi, @mbolivar-nordic, The idea is that these two properties can be used in any node which has "reg" property, currently "reg" is is base.yaml, so I think we'd better put these two together with "reg", otherwise if they are put in some standalone yaml file, for example, put them into dts/bindings/base/zephyr,memory-region.yaml where current "zephyr, memory-region" locates, we have to include this ymal file in each device driver node yaml files which want to use this feature, there will be more modifications.

**Comment by mbolivar-nordic** [comment]
OK, I'm going to leave my -1 on until @galak has a chance to come back to this since he has misgivings as well. I will let you sort it out with him. I have no MMU experience (I'm a microcontroller and userspace sort of engineer!), so it sounds best for me to avoid making this discussion noisier.

**Comment by JiafeiPan** [comment]
> Need to read through the thread, but this really feels like it should be in the driver and NOT in devicetree.

Hi, @galak , not sure whether have go through this thread, and what's your comments? thanks.

**Comment by carlocaione** [comment]
@JiafeiPan if you feel it is valuable you can bring this to the next dev review meeting.

**Comment by galak** [comment]
I'm against this change.  As @carlocaione stated, this should be handled via the MMIO api and having drivers doing the mapping.  Introducing this properties in the dts is NOT the right solution.

**Review by galak — CHANGES_REQUESTED** [review]
I'm against this change.  As @carlocaione stated, this should be handled via the MMIO api and having drivers doing the mapping.  Introducing this properties in the dts is NOT the right solution.

**Comment by JiafeiPan** [comment]
Sorry for the late response because of holiday.

Yes, I agree that for native Zephyr driver, it should handle memory mapping.

Maybe this new feature is only useful for the use-case that application call hal driver APIs directly and don't use native Zephyr drivers, and hal driver mostly likely don't handle memory mapping.

Anyway, I keep my opinions and depend on maintainers decision whether to merge it or not, thanks.

**Pull request closed without merging**
